### ISSUE TITLE
Fix FP S2201 (`no-ignored-return`): "Return values from functions without side effects should not be ignored" triggering on `Array#find`

### ISSUE THREAD

**Issue by thw0rted** [SonarSource/SonarJS#1948]
**I want to report a bug.**

**SonarJS version:** 5.0 (though I think it also applies to 6.2)

**Rule key:** RSPEC-2201

**Reproducer** 
````js
let found;
const haystack = {
  a: [{name, val}, {name, val}],
  b: [{name, val}, {name, val}],
}; // etc
Object.entries(haystack).find(([key, arr]) => {
  found = arr.find(obj => obj.name === needle);
  return !!found;
});
if (found) { ... }
````

**Expected behavior**

The rule shouldn't trigger when the callback passed to `find` has side effects, and assigning to a variable scoped outside of the callback is clearly a side effect.

This is a well-established pattern for iterating over deep / complex structures, and uses the early-exit behavior of `find` to avoid continuing to scan once a hit has been found.

**Comment by vilchik-elena** [comment]
Hi,

Your code could be refactored to avoid changing variable from outer scope:
```
const found = Object.entries(haystack).find(([key, arr]) => {
  return arr.find(obj => obj.name === needle);
});
if (found !== undefined) { ... }
```

I believe that changing var from outer scope makes code harder to follow and maintain

**Comment by thw0rted** [comment]
Your refactor does not have the same behavior as the example code.  `Array#find` returns the value of the array instance that causes the predicate function to return true.  That means that your code will return one tuple (`[key, arr]`) where `arr` is the array containing the record I'm looking for.  I would then have to search the array a second time to actually get the value.  As far as I have been able to figure out, this kind of second-order search with `Array#find` is impossible without side effects.

**Comment by thw0rted** [comment]
Hello from the future!  I am still having this issue using SonarJS v7.2.  Is there anything I can do to help improve the rule?

**Comment by ilia-kebets-sonarsource** [comment]
Idea: we could check in the [`hasSideEffect()`](https://github.com/SonarSource/eslint-plugin-sonarjs/blob/7d1e99353199f22263a63d3022e7a42d08b6daaf/src/rules/no-ignored-return.ts#L254C4-L261) function if there are assignments to variables that are not declared in the function scope.

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-139
